sdks/java/core/src/main/java/org/apache/beam/sdk/values/TypeParameter.java

```diff
@@ -0,0 +1,45 @@
+package org.apache.beam.sdk.values;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+
+/**
+ * Captures a free type variable that can be used in {@link TypeDescriptor#where}. For example:
+ *
+ * <pre>   {@code
+ *   static <T> TypeDescriptor<List<T>> listOf(Class<T> elementType) {
+ *     return new TypeDescriptor<List<T>>() {}
+ *         .where(new TypeParameter<T>() {}, elementType);
+ *   }}</pre>
+ */
+public abstract class TypeParameter<T> {
+  final TypeVariable<?> typeVariable;
+
+  public TypeParameter() {
+    Type superclass = getClass().getGenericSuperclass();
```

Comment:
I'd like to check my understanding here. I have some ideas for giving more verbose error messages about usage. I am happy to implement them myself, but want to see if I have things correct.

Assuming I have a free type variable `X` If I write

    new TypeParameter<X>() {}

then `getClass()` is the anonymous subclass, and `getClass().getGenericSuperclass()` is `TypeParameter<X>`

So can the `check` be even more specific, checking that it is, in fact just a `TypeParameter`? This would allow a more specific error message about the proper usage.

And then if I write

    TypeVariable<String>() {}

I think I'll get a cast exception. Perhaps another check that the value from `getActualTypeArguments()` is a `TypeVariable`, with a friendly message about how it shouldn't be a concrete type.